### PR TITLE
Add support for WP-API error parsing

### LINKED ISSUE #8030: Update DotComError parser to support v2 errors
The v2 version of the WordPress.com API uses the core WordPress REST API, which has a different error format.

We need a new error parser so we can identify errors in the In-app Purchases API requests

### PR DESCRIPTION
**Pull request by koke**
Closes: #8030 

### Description

The existing networking code uses `DotcomValidator` and `DotcomError` to attempt to extract error information from WordPress.com API responses (`error`/`message`). However, the newer endpoints that use the core REST API (`wpcom/v2`, `wp/v2`) use a different error format (`code`/`message`/`data`).

This PR got a bit complex, but the main idea is that each type of request can define a validator that might convert a response into an error. This is useful because a `DotcomRequest` might want to do a different thing depending on the `wordpressApiVersion`.

### Testing instructions

The easiest way to force an error is to modify `InAppPurchasesRemote` and make sure the `X-APP-ID` header has an unknown identifier (e.g. `Bundle.main.bundleIdentifier?.appending(".unknown")`).

Then:

1. Go to Hub menu > Debug IAP
2. Wait for the products to load
3. Look at the console for an error

If you test this on trunk, all you would see is a parsing error:

> Error loading products: typeMismatch(Swift.Array<Any>, Swift.DecodingError.Context(codingPath: [], debugDescription: "Expected to decode Array<Any> but found a dictionary instead.", underlyingError: nil))

With the changes in this PR, you should see a more specific error:

> Error loading products: WordPress API Error: [rest_forbidden] Sorry, you are not allowed to do that.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Model/WordPressApiError.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import Foundation
+
+/// WordPress API Request Error
+///
+public enum WordPressApiError: Error, Decodable, Equatable {
```
Comment (toupper) [inline Networking/Networking/Model/WordPressApiError.swift:5]: We already have an enum for WordPress.com Request Errors (`DotcomError`), for me to understand, do these two enums have different purposes?

**Comment by toupper** [comment]
The code looks good to me and it tests well! 🎉 :shipit: It's nice that we have now more specific error messages:

```
2022-11-04 12:57:32.026167+0100 WooCommerce[12468:3206818] <> Mapping Error: WordPress API Error: [rest_forbidden] Sorry, you are not allowed to do that.
2022-11-04 12:57:32.026460+0100 WooCommerce[12468:3206819] [💰IAP Store] Failed obtaining product list from StoreKit: WordPress API Error: [rest_forbidden] Sorry, you are not allowed to do that.
Error loading products: WordPress API Error: [rest_forbidden] Sorry, you are not allowed to do that.
```
I got three repeated error messages, but I guess it's because we log them at every level of our IAP architecture.

**Review by toupper — APPROVED** [review]

File: Networking/Networking/Model/WordPressApiError.swift (hunk 1 above)
Comment (koke) [reply Networking/Networking/Model/WordPressApiError.swift:5]: This is for the core API errors, which might come from WordPress.com (`wpcom/v2`, `wp/v2`) or directly from a self-hosted site (via `WordPressOrgRequest`). There might be some overlap, but error codes might not necessarily match.

**Comment by koke** [comment]
> I got three repeated error messages

Oh good catch. I think the "Mapping error" shouldn't be there. Previously it would only log if the error was coming from the mapper, but since I consolidated the error handling, it's too verbose now.

**Pull request merged**